GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs

```diff
@@ -1257,17 +1273,107 @@ private string GetNewProjectedShaForPlaceholder(string path)
             return null;
         }
 
+        private void ReExpandFolder(
+            BlobSizes.BlobSizesConnection blobSizesConnection,
+            string relativeFolderPath,
+            ConcurrentDictionary<string, PlaceholderListDatabase.PlaceholderData> updatedPlaceholderList,
+            HashSet<string> existingFolderPlaceholders,
+            ConcurrentHashSet<string> folderPlaceholdersToKeep)
+        {
+            FolderData folderData;
+            if (!this.TryGetOrAddFolderDataFromCache(relativeFolderPath, out folderData))
+            {
+                // Folder is no longer in the projection
+                return;
+            }
+
+            folderData.PopulateSizes(
+                this.context.Tracer,
+                this.gitObjects,
+                blobSizesConnection,
+                availableSizes: null,
+                cancellationToken: CancellationToken.None);
+
+            for (int i = 0; i < folderData.ChildEntries.Count; i++)
+            {
+                FolderEntryData childEntry = folderData.ChildEntries[i];
+                string childRelativePath;
+                if (relativeFolderPath.Length == 0)
+                {
+                    childRelativePath = childEntry.Name.GetString();
+                }
+                else
+                {
+                    childRelativePath = relativeFolderPath + Path.DirectorySeparatorChar + childEntry.Name.GetString();                    
+                }
+
+                if (childEntry.IsFolder)
+                {
+                    if (!existingFolderPlaceholders.Contains(childRelativePath))
+                    {
+                        // TODO(Mac): Issue #245, handle failures of WritePlaceholder
+                        this.fileSystemVirtualizer.WritePlaceholder(
+                            childRelativePath,
+                            endOfFile: 0, 
+                            isDirectory: true, 
+                            shaContentId: GVFSConstants.AllZeroSha);
+
+                        updatedPlaceholderList.TryAdd(
+                            childRelativePath,
+                            new PlaceholderListDatabase.PlaceholderData(childRelativePath, GVFSConstants.AllZeroSha));
+
+                        folderPlaceholdersToKeep.Add(relativeFolderPath);
+                    }
+                }
+                else
+                {
+                    if (!updatedPlaceholderList.ContainsKey(childRelativePath))
+                    {
+                        FileData childFileData = childEntry as FileData;
+                        string sha = childFileData.Sha.ToString();
+
+                        // TODO(Mac): Issue #245, handle failures of WritePlaceholder
+                        this.fileSystemVirtualizer.WritePlaceholder(childRelativePath, childFileData.Size, isDirectory: false, shaContentId: sha);
```

Comment:
Break out the parameters onto separate lines to match the style above?

Reply:
And actually the contents of this if block and the one above are virtually identical. Is there an easy way to avoid the duplication?

Reply:
I'll work on cleaning this up